### PR TITLE
Mark order notification read if needed on watch app and notification extension

### LINKED ISSUE #13873: Mark order notification read if needed on watch app and notification extension
As a continuation on https://github.com/woocommerce/woocommerce-ios/pull/13734 we need to add marking order notification as read on Watch app and in notification extension in a same logic as we do in the app.
Because it needs a bit more code changes we did not do it in https://github.com/woocommerce/woocommerce-ios/pull/13734 but will do in separate PR(s).

### PR DESCRIPTION
**Pull request by bozidarsevo**
Closes: #13873

## Description

As a continuation on https://github.com/woocommerce/woocommerce-ios/pull/13734 this PR adds marking order notification as read on Watch app and in notification extension in a same logic as we do in the app, using the same endpoints but using `NotificationsRemote` directly instead of `NotificationAction` since `NotificationAction` is not available on the watch.

Updates for MarkOrderAsReadUseCase to have two versions of markOrderNoteAsReadIfNeeded, one usses `NotificationsRemote` and the second one uses `NotificationAction`. That way we can have same functionality in the iOS app, notification extension and watchOS app.

Good question would be why not just use `NotificationsRemote` in all places. When we use `NotificationAction` we can see that `NotificationStore` in `synchronizeNotification()` function also calls `updateLocalNotes(with: notes)` and that is a useful addition worth having 2 implementations.

## Testing information

General notification opening for the watchOS app and notification extension
- Do this testing on both watch app and notification extension. When testing the notification extension press the notification and check the notification preview is shown correctly.
- Get few new orders for one of the stores you have so it triggers notifications for it. It is important to have more of them so you can open order that is not the last.
- Tap on the notification for the order that was not the last order
- Check that the app opens the proper order
- Tap on the notification for the last order
- Check that the app opens the proper order and that the notification is marked as read also on the web

https://github.com/user-attachments/assets/80865274-ad87-4cee-a45b-ea1f198e3517

IMG_723E898202D3-1

## Screenshots

There are no user-facing/UI changes but screenshots from https://github.com/woocommerce/woocommerce-ios/pull/13734 can be looked as a reference.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/223">20.4</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/Model/MarkOrderAsReadUseCase.swift [hunk 1]
```diff
@@ -1,30 +1,79 @@
...
+        case updateReadStatus(Swift.Error)
+        case unavailableNote
+    }
     /// Method that marks the order note as read if it is the notification for the last order.
     /// We do it in a way that first we syncronize notification to get the remote `Note`
     /// and then we compare local `orderID` with the one from remote `Note`.
     /// If they match we mark it as read.
     /// We pass syncronized note and error in the `onCompletion` completion block
+#if canImport(Yosemite)
     static func markOrderNoteAsReadIfNeeded(stores: StoresManager, noteID: Int64, orderID: Int, onCompletion: ((Note?, Error?) -> Void)? = nil) {
```
Comment (bozidarsevo) [inline WooCommerce/Classes/Model/MarkOrderAsReadUseCase.swift]: Updates for **MarkOrderAsReadUseCase** to have two versions of **markOrderNoteAsReadIfNeeded**, one usses **NotificationsRemote** and the second one uses **NotificationAction**. That way we can have same functionality in the iOS app, notification extension and watchOS app.

Good question would be why not just use **NotificationsRemote** in all places. When we use **NotificationAction** we can see that **NotificationStore** in **synchronizeNotification**() function also calls **updateLocalNotes**(with: notes) and that is a useful addition worth having 2 implementations.

File: WooCommerce/Classes/Model/MarkOrderAsReadUseCase.swift [hunk 2]
```diff
@@ -1,30 +1,79 @@
...
         }
         stores.dispatch(action)
     }
+#endif
+    /// Method that marks the order note as read if it is the notification for the last order.
+    /// We do it in a way that first we syncronize notification to get the remote `Note`
+    /// and then we compare local `orderID` with the one from remote `Note`.
+    /// If they match we mark it as read.
+    /// We pass syncronized note id and error in the `onCompletion` completion block
+    static func markOrderNoteAsReadIfNeeded(network: Network, noteID: Int64, orderID: Int, onCompletion: ((Int64?, Error?) -> Void)? = nil) {
```
Comment (Ecarrion) [inline WooCommerce/Classes/Model/MarkOrderAsReadUseCase.swift]: Do you think we should take the time and improve this method signature to return a result rather than a tuple?

Comment (Ecarrion) [reply WooCommerce/Classes/Model/MarkOrderAsReadUseCase.swift]: Or even better, we could make it an `async` function so it can be used seamlessly from the notification/watch target.

Updating the function above will be great too!

File: WooCommerce/NotificationExtension/OrderNotificationDataService.swift [hunk 3]
```diff
@@ -96,6 +96,7 @@ final class OrderNotificationDataService {
             ordersRemote.loadOrder(for: Int64(siteID), orderID: Int64(orderID)) { order, error in
                 switch (order, error) {
                 case (let order?, nil):
+                    MarkOrderAsReadUseCase.markOrderNoteAsReadIfNeeded(network: self.network, noteID: notification.noteID, orderID: orderID)
```
Comment (Ecarrion) [inline WooCommerce/NotificationExtension/OrderNotificationDataService.swift]: I don't think we should mark the notification as read when `quick previewing` the notification on the phone.
I think we should only do it when presenting the order detail in both the app (already done) and in the watch (this update).

And if we only do it on the watch, then it is probably safer to do when presenting the order on `OrderDetailLoaderViewModel/OrderDetailLoaderViewModel` 

What do you think?

Comment (bozidarsevo) [reply WooCommerce/NotificationExtension/OrderNotificationDataService.swift]: Hm, could point! I was also having some dilemmas where to call this for this PR, was not sure 100% where to put it. But you have a good point!

File: WooCommerce/Woo Watch App/Orders/OrderDetailLoader.swift [hunk 4]
```diff
@@ -21,6 +26,9 @@ struct OrderDetailLoader: View {
                     .redacted(reason: .placeholder)
             case .loaded(let order):
                 OrderDetailView(order: order)
+                    .onAppear {
```
Comment (bozidarsevo) [inline WooCommerce/Woo Watch App/Orders/OrderDetailLoader.swift]: I was thinking if I should use `.onAppear` here to call `MarkOrderAsReadUseCase.markOrderNoteAsReadIfNeeded` or call it directly in `OrderDetailView`, but I went with this just so we do this marking logic when we do the order fetching from the push notification and not always when we show the order from the watch app.
What do you think?

File: WooCommerce/Classes/Model/MarkOrderAsReadUseCase.swift (hunk 2 above)
Comment (bozidarsevo) [reply WooCommerce/Classes/Model/MarkOrderAsReadUseCase.swift]: On it! 👍

Comment (bozidarsevo) [reply WooCommerce/Classes/Model/MarkOrderAsReadUseCase.swift]: Done! Now the method line number is a bit bigger but hopefully it is ok :)

File: WooCommerce/Woo Watch App/Orders/OrderDetailLoader.swift (hunk 4 above)
Comment (Ecarrion) [reply WooCommerce/Woo Watch App/Orders/OrderDetailLoader.swift]: I think a better place to not add extra logic to the UI and keep it containable in the VM would be:

- Adda a `markNotification...` method in the `OrderNotificationDataService` which will call your use case.
- Add a `markNotification...` method in the `OrderDetailLoaderViewModel` and call the method from the data service
- Call that VM method from the `OrderDetailLoaderView`, it could be on a `task` modifier so it not called multiple times (possibly)

What do you think?

Comment (bozidarsevo) [reply WooCommerce/Woo Watch App/Orders/OrderDetailLoader.swift]: > * Call that VM method from the `OrderDetailLoaderView`

you mean `OrderDetailView`?

Comment (bozidarsevo) [reply WooCommerce/Woo Watch App/Orders/OrderDetailLoader.swift]: Since `OrderDetailView` is used both in `OrderDetailLoader` and `OrdersListView` I was thinking of just having that marking in `OrderDetailLoader` since there we do the loading related to the push notification.

Also I wanted to avoid adding marking logic in `OrderNotificationDataService` since that is used for load/fetching objects and not presenting the order, and we need to mark the notification read if the order is shown, not just loaded/fetched.

I would maybe just change this to:
```
OrderDetailView(order: order)
          .task {
            await MarkOrderAsReadUseCase.markOrderNoteAsReadIfNeeded(network: self.network,
                                         noteID: pushNotification.noteID,
                                         orderID: Int(order.orderID))
          }
```

In short:
- in the app we will have marking in `OrderLoaderViewController` (`markNotificationAsReadIfNeeded(note: Note)`) which is called if users just tap the notification or enter the app via notification extension.
- on watch we have marking in `OrderDetailLoader` when we present `OrderDetailView` with loaded order

Comment (Ecarrion) [reply WooCommerce/Woo Watch App/Orders/OrderDetailLoader.swift]: > Also I wanted to avoid adding marking logic in OrderNotificationDataService since that is used for load/fetching objects and not presenting the order,

I think `OrderNotificationDataService` is fine to be updated, the name states that notifications is part of its scope.

> I would maybe just change this to:

```
OrderDetailView(order: order)
    .task {
        await MarkOrderAsReadUseCase.markOrderNoteAsReadIfNeeded(network: self.network, noteID pushNotification.noteID, orderID: Int(order.orderID))
     }
``` 

I think this will work fine,  what I want to avoid is creating multiple instances of `network` for the same module because this increases the possibilities that they grow apart in the future, for example applying one configuration to the one in the view model and forgetting to update the one in the view.

It also makes it easier to be updated in the future, as the VM has all the state about the recent push notification.

Does that make sense?

Comment (bozidarsevo) [reply WooCommerce/Woo Watch App/Orders/OrderDetailLoader.swift]: I understand! Will try to update it now 👍

**Review by Ecarrion — APPROVED** [review]
Looks and tests well @bozidarsevo! Thanks for bearing with me during the review 😇 

There was only an unsed import and I took the liberty to remove it in https://github.com/woocommerce/woocommerce-ios/pull/13896/commits/8b5fe12c61265612734756457021ccbd2556caec in order to not delay you further.

I hope you don't mind :-)

**Comment by bozidarsevo** [comment]
> I hope you don't mind :-)

Not at all! Thanks ;)
